File: Sources/AppSuppression/DefaultExclusions.swift
import Foundation

/// What the app excludes out of the box.
///
/// Two lists, because bundle identifiers cannot be written from memory any more
/// safely than Unicode codepoints can. A wrong identifier fails silently and
/// the first thing the user notices is a corrupted document, so:
///
/// - `bundleIDs` holds identifiers verified against installed copies of the
///   apps, or well enough established to be safe.
/// - `applicationNames` is resolved against what is actually on the disk at
///   launch, reading each app's real identifier out of its bundle. This is how
///   the TeX editors get covered: their identifiers are not ones to guess at,
///   and the answer is sitting in `/Applications` waiting to be read.
///
/// Both lists only ever produce rules for apps that are installed, so the
/// exclusion editor shows a short, true list rather than a wall of software the
/// user does not have.
public enum DefaultExclusions {

    public static let bundleIDs: [String] = [
        // Terminals — a shell is nothing but backslashes.
        "com.apple.Terminal",
        "com.googlecode.iterm2",
        "net.kovidgoyal.kitty",
        "com.mitchellh.ghostty",
        "dev.warp.Warp-Stable",
        "com.github.wez.wezterm",
        "co.zeit.hyper",

        // Code editors — where a .tex file is most likely to be open.
        "com.microsoft.VSCode",
        "com.microsoft.VSCodeInsiders",
        "com.todesktop.230313mzl4w4u92",   // Cursor
        "com.apple.dt.Xcode",
        "com.sublimetext.4",
        "com.sublimetext.3",
        "com.barebones.bbedit",
        "org.gnu.Emacs",
        "org.vim.MacVim",
        "dev.zed.Zed",
        "com.panic.Nova",
        "com.rstudio.desktop",             // Sweave and R Markdown are LaTeX
    ]

    /// Matched against `CFBundleName` and the file name of each installed app.
    ///
    /// The TeX editors are here rather than above because their identifiers are
    /// exactly the kind of thing that looks obvious and is wrong.
    public static let applicationNames: [String] = [
        "TeXShop",
        "TeXstudio",
        "Texifier",
        "Texpad",
        "TeXmaker",
        "Texmaker",
        "TeXworks",
        "TeXnicCenter",
        "LyX",
        "Overleaf",

        // Repeated from `bundleIDs` on purpose: if an identifier up there is
        // wrong, the name catches the app anyway.
        "Visual Studio Code",
        "VSCodium",
        "Cursor",
        "Sublime Text",
        "Zed",
        "Nova",
        "CotEditor",
        "MacVim",
        "Emacs",
        "BBEdit",
        "RStudio",
    ]

    /// Overleaf is the case the whole phase exists for: it is not an app, it is
    /// a tab in a browser the user also chats in.
    public static let sites: [ExcludedSite] = [
        ExcludedSite(host: "overleaf.com",
                     titleFragments: ["Overleaf"]),
        ExcludedSite(host: "sharelatex.com",
                     titleFragments: ["ShareLaTeX"]),
        ExcludedSite(host: "papeeria.com",
                     titleFragments: ["Papeeria"]),
        ExcludedSite(host: "cocalc.com",
                     titleFragments: ["CoCalc"]),
        ExcludedSite(host: "latexbase.com",
                     titleFragments: ["LaTeX Base"]),
    ]
}

File: windows/LaTeXSquiggly.Core/Suppression/DefaultExclusions.cs
namespace LaTeXSquiggly.Core.Suppression;

/// <summary>
/// What the app excludes out of the box.
///
/// The Mac version resolves TeX editors off the disk because bundle identifiers
/// cannot be written from memory safely. Windows is kinder here: an app is
/// identified by its executable name, which is exactly what the user sees in
/// Task Manager and what an installer cannot quietly change. So the list is
/// literal, and a name that matches nothing installed simply never fires.
///
/// The rule behind every entry: an app that corrupts LaTeX source is worse than
/// no app at all, so anywhere LaTeX source is plausibly written, stay quiet.
/// </summary>
public static class DefaultExclusions
{
    /// <summary>Executable names, lowercased, without .exe.</summary>
    public static readonly (string Process, string Name)[] Apps =
    {
        // TeX editors. The reason this list exists.
        ("texstudio", "TeXstudio"),
        ("texmaker", "Texmaker"),
        ("texworks", "TeXworks"),
        ("miktex-texworks", "TeXworks"),
        ("texniccenter", "TeXnicCenter"),
        ("winedt", "WinEdt"),
        ("lyx", "LyX"),
        ("texifier", "Texifier"),
        ("texpad", "Texpad"),
        ("scientificworkplace", "Scientific WorkPlace"),

        // Terminals. A shell is nothing but backslashes.
        ("windowsterminal", "Windows Terminal"),
        ("cmd", "Command Prompt"),
        ("powershell", "Windows PowerShell"),
        ("pwsh", "PowerShell"),
        ("conhost", "Console Window Host"),
        ("mintty", "Git Bash"),
        ("bash", "Bash"),
        ("alacritty", "Alacritty"),
        ("wezterm-gui", "WezTerm"),
        ("putty", "PuTTY"),
        ("hyper", "Hyper"),

        // Code editors, where a .tex file is most likely to be open.
        ("code", "Visual Studio Code"),
        ("code - insiders", "Visual Studio Code Insiders"),
        ("codium", "VSCodium"),
        ("cursor", "Cursor"),
        ("windsurf", "Windsurf"),
        ("zed", "Zed"),
        ("sublime_text", "Sublime Text"),
        ("notepad++", "Notepad++"),
        ("gvim", "gVim"),
        ("vim", "Vim"),
        ("nvim", "Neovim"),
        ("emacs", "Emacs"),
        ("runemacs", "Emacs"),
        ("devenv", "Visual Studio"),
        ("rstudio", "RStudio"),          // Sweave and R Markdown are LaTeX
        ("idea64", "IntelliJ IDEA"),
        ("pycharm64", "PyCharm"),
        ("clion64", "CLion"),
        ("webstorm64", "WebStorm"),
        ("rider64", "Rider"),
        ("jupyter-notebook", "Jupyter"),
        ("obsidian", "Obsidian"),        // its maths blocks are LaTeX source
        ("typora", "Typora"),
    };

    /// <summary>
    /// Overleaf is the case the whole feature exists for: it is not an app, it
    /// is a tab in a browser the user also chats in.
    ///
    /// The title fragments do the work on Windows, where a browser gives us its
    /// window title rather than its URL.
    /// </summary>
    public static ExcludedSite[] Sites() => new[]
    {
        new ExcludedSite("overleaf.com", "Overleaf"),
        new ExcludedSite("sharelatex.com", "ShareLaTeX"),
        new ExcludedSite("papeeria.com", "Papeeria"),
        new ExcludedSite("cocalc.com", "CoCalc"),
        new ExcludedSite("latexbase.com", "LaTeX Base"),
    };

    public static ExclusionList Fresh()
    {
        var list = new ExclusionList();
        foreach (var (process, name) in Apps) list.Add(new ExcludedApp(process, name));
        foreach (var site in Sites()) list.Add(site);
        return list;
    }
}
